### PR TITLE
Add LLM improvements

### PR DESCRIPTION
**Pull request by RichDom2185**
- [ ] Access control
- [ ] Logging & auditing
- [x] Move prompt generation to BE to prevent bypass

### CODE REVIEW

**Comment by RichDom2185** [comment]
@yiwen101 the prompt builder is separated from SICP notes so that we can put AG logic inside the builder and retain the abstraction

File: lib/cadet_web/plug/rate_limiter.ex [hunk 1]
```diff
@@ -0,0 +1,24 @@
+defmodule CadetWeb.Plugs.RateLimiter do
+  import Plug.Conn
+
+  @rate_limit 500
```
Comment (RichDom2185) [inline lib/cadet_web/plug/rate_limiter.ex:5]: Does this mean 500 requests every 24 hours?

**Review by yiwen101 — APPROVED** [review]
LGTM

**Comment by RichDom2185** [comment]
Corresponding FE PR: https://github.com/source-academy/frontend/pull/3010

**Pull request merged**